polkadot/node/network/collator-protocol/src/validator_side_experimental/peer_manager/mod.rs

```diff
@@ -0,0 +1,495 @@
+			scheduled_paras.contains(p)
+		}) {
+			// The new set is a superset of the old paras and their lengths are equal, so they are
+			// identical.
+
+			if prev_paras_count == scheduled_paras.len() {
+				// Nothing to do if the scheduled paras didn't change.
+				return
+			}
+		}
+
+		// Borrow checker can't tell that prev_scheduled_paras is not used anymore.
+		std::mem::drop(prev_scheduled_paras);
+
+		// Recreate the connected peers based on the new schedule and try populating it again based
+		// on their reputations. Disconnect any peers that couldn't be kept
+		let mut new_instance =
+			ConnectedPeers::new(scheduled_paras, CONNECTED_PEERS_LIMIT, CONNECTED_PEERS_PARA_LIMIT);
+
+		std::mem::swap(&mut new_instance, &mut self.connected);
+		let prev_instance = new_instance;
+		let (prev_peers, cached_scores) = prev_instance.consume();
+
+		// Build a closure that can be used to first query the in-memory past reputations of the
+		// peers before reaching for the DB.
+		// TODO: we could warm-up the DB for these specific paraids. (Cache them in memory on the
+		// Backend impl)
+
+		// Borrow these for use in the closure.
+		let cached_scores = &cached_scores;
+		let db = &self.db;
+		let reputation_query_fn = |peer_id: PeerId, para_id: ParaId| async move {
+			if let Some(cached_score) =
+				cached_scores.get(&para_id).and_then(|per_para| per_para.get_score(&peer_id))
+			{
+				cached_score
+			} else {
+				db.query(&peer_id, &para_id).await.unwrap_or_default()
+			}
+		};
+
+		// See which of the old peers we should keep.
+		let mut peers_to_disconnect = vec![];
+		for (peer_id, peer_info) in prev_peers {
+			let outcome = self.connected.try_accept(reputation_query_fn, peer_id, peer_info).await;
+
+			match outcome {
+				TryAcceptOutcome::Rejected => {
+					peers_to_disconnect.push(peer_id);
+				},
+				TryAcceptOutcome::Replaced(replaced_peer_ids) => {
+					peers_to_disconnect.extend(replaced_peer_ids);
+				},
+				TryAcceptOutcome::Added => {},
+			}
+		}
+
+		// Disconnect peers that couldn't be kept.
+		self.disconnect_peers(sender, peers_to_disconnect).await;
+	}
+
+	/// Process a declaration message of a peer.
+	pub async fn declared<Sender: CollatorProtocolSenderTrait>(
+		&mut self,
+		sender: &mut Sender,
+		peer_id: PeerId,
+		para_id: ParaId,
+	) {
+		let outcome = self.connected.declared(peer_id, para_id);
+
+		match outcome {
+			DeclarationOutcome::Accepted => {
+				gum::debug!(
+					target: LOG_TARGET,
+					?para_id,
+					?peer_id,
+					"Peer declared",
+				);
+			},
+			DeclarationOutcome::Switched(old_para_id) => {
+				gum::debug!(
+					target: LOG_TARGET,
+					?para_id,
+					?old_para_id,
+					?peer_id,
+					"Peer switched collating paraid",
+				);
+			},
+			DeclarationOutcome::Rejected => {
+				gum::debug!(
+					target: LOG_TARGET,
+					?para_id,
+					?peer_id,
+					"Peer declared but rejected. Going to disconnect.",
+				);
+
+				self.disconnect_peers(sender, vec![peer_id]).await;
```

Comment:
Premature optimization alert: I wonder whether it would make sense to make the limit "soft" and not send this message directly, but collect a few first in &mut self and only send this message once we reached e.g. 10 or more nodes.

Reply:
Take a look at the reputation aggregator, it can probably be made more generic to aggregate this kind of requests as well. But in a very future PR :)